pkg/helmreconciler/listeners.go

```diff
@@ -0,0 +1,472 @@
+}
+
+// BeginChart delegates BeginChart to the Listeners in first to last order.
+func (l *CompositeRenderingListener) BeginChart(chart string, manifests []manifest.Manifest) ([]manifest.Manifest, error) {
+	allErrors := make([]error, 0, len(l.Listeners))
+	var err error
+	for _, listener := range l.Listeners {
+		if manifests, err = listener.BeginChart(chart, manifests); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return manifests, utilerrors.NewAggregate(allErrors)
+}
+
+// BeginResource delegates BeginResource to the Listeners in first to last order.
+func (l *CompositeRenderingListener) BeginResource(obj runtime.Object) (runtime.Object, error) {
+	allErrors := make([]error, 0, len(l.Listeners))
+	var err error
+	for _, listener := range l.Listeners {
+		if obj, err = listener.BeginResource(obj); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return obj, utilerrors.NewAggregate(allErrors)
+}
+
+// ResourceCreated delegates ResourceCreated to the Listeners in first to last order.
+func (l *CompositeRenderingListener) ResourceCreated(created runtime.Object) error {
+	allErrors := make([]error, 0, len(l.Listeners))
+	for _, listener := range l.Listeners {
+		if err := listener.ResourceCreated(created); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// ResourceUpdated delegates ResourceUpdated to the Listeners in first to last order.
+func (l *CompositeRenderingListener) ResourceUpdated(updated runtime.Object, old runtime.Object) error {
+	allErrors := make([]error, 0, len(l.Listeners))
+	for _, listener := range l.Listeners {
+		if err := listener.ResourceUpdated(updated, old); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// ResourceDeleted delegates ResourceDeleted to the Listeners in first to last order.
+func (l *CompositeRenderingListener) ResourceDeleted(deleted runtime.Object) error {
+	allErrors := make([]error, 0, len(l.Listeners))
+	for _, listener := range l.Listeners {
+		if err := listener.ResourceDeleted(deleted); err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// ResourceError delegates ResourceError to the Listeners in first to last order.
+func (l *CompositeRenderingListener) ResourceError(obj runtime.Object, err error) error {
+	// reverse order for completions
+	allErrors := make([]error, 0, len(l.Listeners))
+	for index := len(l.Listeners) - 1; index > -1; index-- {
+		if listenerErr := l.Listeners[index].ResourceError(obj, err); listenerErr != nil {
+			allErrors = append(allErrors, listenerErr)
+		}
+	}
+	return utilerrors.NewAggregate(allErrors)
+}
+
+// EndResource delegates EndResource to the Listeners in last to first order.
+func (l *CompositeRenderingListener) EndResource(obj runtime.Object) error {
+	// reverse order for completions
+	allErrors := make([]error, 0, len(l.Listeners))
```

Comment:
you don't have to make slices for appending (you can append to nil slice).
use var errs []error here instead.

Reply:
This simply prevents resizing the slice.  At most, there will be len(l.Listeners) errors, so we presize the slice to avoid allocating memory during append.

Reply:
But what if you have no errors at all (most common case)? Then you end up allocating and freeing memory every time unnecessarily.

Reply:
A fair point, although I don't see why you didn't mention this in the first place.  Still, this shouldn't be holding up the PR.

Reply:
I didn't mention it because I didn't look at the behavior in detail till you explained why you were pre-sizing. 
You really want to check this in rather than just fix it?